include/eve/arch/riscv/as_register.hpp

```diff
@@ -0,0 +1,497 @@
+//==================================================================================================
+/*
+  EVE - Expressive Vector Engine
+  Copyright : EVE Project Contributors
+  SPDX-License-Identifier: BSL-1.0
+*/
+//==================================================================================================
+#pragma once
+
+#include <eve/arch/riscv/predef.hpp>
+#include <eve/traits/as_integer.hpp>
+
+#include <type_traits>
+#if defined(EVE_INCLUDE_RISCV_HEADER)
+namespace eve
+{
+template<typename T> struct logical;
+}
+
+namespace eve
+{
+template<typename Type, typename Size, rvv_abi ABI> struct as_register<Type, Size, ABI>
```

Comment:
I think, now that I understand it, I can simplify this code a bit. 
However, I would like to compile the code I'd suggest first.

Is there a way I can write rvv simd code on godbolt?

Reply:
I suppose you can. Here is an example: https://godbolt.org/z/ThfEqhc59  (just vector add with lmul=1 and scalar type `std::uint64_t`). The main idea here is to add to riscv clang `-march=rv64gcv -mrvv-vector-bits=128`.

By the way, here I use load with mask, as in EVE, but actually I can perform nomask load. This should be faster, but will result in much more code for review (as I will need to write separate intrinsic for each combination of type/lmul). I suggest to make this optimization in separate review.

Reply:
Here is what I have: https://godbolt.org/z/GcEz5v95j
Feel vindicated, I had no idea that they decided to support none of the normal c++ things, we have to some macro.

I think having one long table is nice for me.

Reply:
I've used some defines from your code. Unfortunately, your calculation of LMUL not type-dependent, so it can not be used ( as some types, like `int`, can not be used with frac lmul 1/4 or 1/8 for example), so I still use the one that is placed in tags (I need the these calculations for `logical` too, so placing it in `as_register` is not convenient). 

> one long table is nice for me

I still prefer to use tree-like function calls, so, if you do not mind, I prefer to keep for now structure of getting registers, as set of short functions looks better (and will be easier to debug) in my opinion. If you think that one long function will be better, I will replace this.

By the way, in the original review (before deleting all the macros), I used more interesting macros, with concatenation, so you will not need to use for example `4` (as LMUL value) several times (and `4`). May be return that style, as we already return defines?

Reply:
> Uortunately, your calculation of LMUL not type-dependent, so it can not be use so it can not be used
Yes, it is: 

```
  constexpr std::ptrdiff_t m1_len = __riscv_v_fixed_vlen;
  constexpr std::ptrdiff_t min_len = m1_len * sizeof(T) / 8;
  constexpr std::ptrdiff_t reg_len = std::max(min_len, N);
```



> set of short functions looks better
I don't agree with you here but I can read it now, so w/e.

> more interesting macros
No, that one is a no

Reply:
Doesn't saw that you use `constexpr std::ptrdiff_t min_len = m1_len * sizeof(T) / 8;`. It seems by this you can get expected frac lmul. I've refactored code to the same manner. Could you take a look? I still need to change code, as `std::max(min_len, N);` looks like error (max between cardinal and vector register size).